Support for the AirThings Wave Plus

jeromelaban opened the pull request:
# What does this implement/fix? 

Adds support for the [AirThings Wave Plus](https://www.airthings.com/wave-plus) BLE device.

## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Configuration change (this will require users to update their yaml configuration files to keep working)

**Related issue or feature (if applicable):** fixes https://github.com/esphome/feature-requests/issues/147

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):** https://github.com/esphome/esphome-docs/pull/1081
  
# Test Environment

- [x] ESP32
- [ ] ESP8266
- [ ] Windows
- [ ] Mac OS
- [ ] Linux

## Example entry for `config.yaml`:
<!--
  Supplying a configuration snippet, makes it easier for a maintainer to test
  your PR. Furthermore, for new integrations, it gives an impression of how
  the configuration would look like.
  Note: Remove this section if this PR does not have an example entry.
-->

BLE Scanner to find the device's MAC:
```yaml
esp32_ble_tracker:
airthings_wave_plus:
```

Sensors:
```yaml
  - platform: airthings_wave_plus
    mac_address: 'xx:xx:xx:xx:xx:xx'
    update_interval: 5min
    temperature:
      name: "Wave Plus Temperature"
    radon:
      name: "Wave Plus Radon"
    radon_long_term:
      name: "Wave Plus Radon Long Term"
    pressure:
      name: "Wave Plus Pressure"
    humidity:
      name: "Wave Plus Humidity"
    co2:
      name: "Wave Plus CO2"
    tvoc:
      name: "Wave Plus VOC"
```

# Explain your changes

This PR is adding a AirThings device scanner, to find the BLE MAC address of the device, then a set of sensors provided by the device, polled every 5 minutes.

This PR adds a class, unit and icon related to radon measurement.

This is a new repository for me, so I'm pretty sure the esphome wizards will find things to say about it and how to make this PR better. 

Tests and documentation are coming.

## Checklist:
  - [x] The code change is tested and works locally.
  - [x] Tests have been added to verify that the new code works (under `tests/` folder).
  
If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

jeromelaban committed:
feat: Add support for radon related classes, unit and icon

jeromelaban committed:
feat: Add airthings ble listener and sensors

jeromelaban committed:
chore: Move listener to its own component

jeromelaban committed:
test: Add airthings_ble, airthings_wave_plus tests

jeromelaban committed:
chore: adjust naming, formatting

jeromelaban commented:
The code is looking better with some formatting applied, though the clang linter does not know how to find the "BLEDevice.h" header file, and requests changes that cannot be made (`onConnect` to `on_connect` for an overriden method).

Is there a way to hint the linter to find this arduino file ?

buxtronix commented:
Could the BLE client part of this be managed by https://github.com/esphome/esphome/pull/1177  ?

jeromelaban commented:
@buxtronix it may end up being the base for the BLE Client, yes, but for now the integration in this PR is built around not keeping the BLE connection alive on purpose, so that the AirThings mobile app continues to work. I'm not sure if #1177 supports this mode yet.

buxtronix commented:
> @buxtronix it may end up being the base for the BLE Client, yes, but for now the integration in this PR is built around not keeping the BLE connection alive on purpose, so that the AirThings mobile app continues to work (I'll update the PR to mention that in the code and docs as well). I'm not sure if #1177 supports this mode yet.

Yep, #1177 supports enabling and disabling individual client connections.

jeromelaban commented:
Got it, it's up to the sensor to enable/disable itself and wait for the connect event to get the attribute value. I don't follow all the pieces of that new API completely, but as long as there's a way to read an attribute raw value, it should be good.

jeromelaban committed:
fix: airwave things plus tests typo

jeromelaban committed:
fix: Validate for out-of-range radon, voc and co2

jeromelaban committed:
chore: add missing update interval support

sylvinus commented:
Hey @jeromelaban, funny seeing you here :)

I just tested this successfully with a 1st gen Wave Plus (Model 2930xxxx).

Only issue I had was https://github.com/esphome/esphome-docs/pull/1081/files#r664574473

Great work!

XBeg9 commented:
I would love to integrate my Airthings Wave Plus using esphome, anything I can help with?

jeromelaban committed:
chore: Update airthings sensor.py update interval

jeromelaban commented:
> Only issue I had was https://github.com/esphome/esphome-docs/pull/1081/files#r664574473
> 
> Great work!

Thanks @sylvinus! Indeed, I missed the sensor.py configuration update for the `update_interval` to be recognized in the configuration. Last update should work better now!

jeromelaban committed:
chore: Remove invalid file

jeromelaban committed:
chore: Apply c++ linting suggestions

jeromelaban committed:
chore: Apply python linting suggestions

OttoWinter commented on esphome/components/airthings_ble/airthings_listener.cpp:
```diff
@@ -0,0 +1,30 @@
+bool AirthingsListener::parse_device(const esp32_ble_tracker::ESPBTDevice &device) {
+  for (auto &it : device.get_manufacturer_datas()) {
+    if (it.uuid == esp32_ble_tracker::ESPBTUUID::from_uint16(0x0334)) {
+      uint SN = it.data[0];
+      SN |= ((uint) it.data[1] << 8);
+      SN |= ((uint) it.data[2] << 16);
+      SN |= ((uint) it.data[3] << 24);
```
```suggestion
      if (it.data.size() < 4)
        continue;
      uint32_t sn = it.data[0];
      sn |= ((uint32_t) it.data[1] << 8);
      sn |= ((uint32_t) it.data[2] << 16);
      sn |= ((uint32_t) it.data[3] << 24);
```

Just adds a check to prevent out-of-bounds reading, and replaces `uint` with the more concrete `uint32_t` here, and fixes variable naming to match the repos settings.

OttoWinter commented on esphome/components/airthings_ble/airthings_listener.cpp:
```diff
@@ -0,0 +1,30 @@
+      SN |= ((uint) it.data[2] << 16);
+      SN |= ((uint) it.data[3] << 24);
+
+      ESP_LOGD(TAG, "Found AirThings device Serial:%u (MAC: %s)", SN, device.address_str().c_str());
```
```suggestion
      ESP_LOGD(TAG, "Found AirThings device Serial:%u (MAC: %s)", sn, device.address_str().c_str());
```

OttoWinter commented on esphome/components/airthings_wave_plus/airthings_wave_plus.cpp:
```diff
@@ -0,0 +1,182 @@
+  last_value_time_ = millis();
+
+  // Schedule to avoid a deadlock
+  set_timeout(0, [this] { this->read_sensors_(); });
```
```suggestion
  defer([this] { this->read_sensors_(); });
```

OttoWinter commented on esphome/components/airthings_wave_plus/airthings_wave_plus.cpp:
```diff
@@ -0,0 +1,182 @@
+  client_->disconnect();
+}
+
+boolean AirthingsWavePlus::is_valid_radon_value_(short radon) { return 0 <= radon && radon <= 16383; }
+
+boolean AirthingsWavePlus::is_valid_voc_value_(short voc) { return 0 <= voc && voc <= 16383; }
+
+boolean AirthingsWavePlus::is_valid_co2_value_(short co2) { return 0 <= co2 && co2 <= 16383; }
```
```suggestion
bool AirthingsWavePlus::is_valid_radon_value_(short radon) { return 0 <= radon && radon <= 16383; }

bool AirthingsWavePlus::is_valid_voc_value_(short voc) { return 0 <= voc && voc <= 16383; }

bool AirthingsWavePlus::is_valid_co2_value_(short co2) { return 0 <= co2 && co2 <= 16383; }
```

OttoWinter commented on esphome/components/airthings_wave_plus/airthings_wave_plus.h:
```diff
@@ -0,0 +1,90 @@
+#pragma once
+
+#include "esphome/core/component.h"
+#include "esphome/core/helpers.h"
+#include "esphome/core/preferences.h"
+#include "esphome/core/log.h"
+#include "esphome/core/application.h"
+#include "esphome/core/version.h"
```
```suggestion
#include "esphome/core/component.h"
#include "esphome/core/helpers.h"
#include "esphome/core/log.h"

#ifdef ARDUINO_ARCH_ESP32
```

- I don't think all those includes are used
- Adds ifdef check so linter doesn't activate because that one uses an ESP8266 toolchain

jeromelaban replied:
I had to keep #include `"esphome/core/application.h"` for `sensor` to be defined.

OttoWinter commented on esphome/components/airthings_wave_plus/airthings_wave_plus.h:
```diff
@@ -0,0 +1,90 @@
+  uint32_t update_count_ = 0;
+  bool connected_ = false;
+  bool connecting_ = false;
+  int32_t last_value_time_;
+  int32_t last_connect_time_ = connection_timeout_in_seconds_ * -1000;
```
```suggestion
  uint32_t last_value_time_;
  uint32_t last_connect_time_;
```

These have to be unsigned, otherwise you'll get issues after 20 days of uptime (when the millis function rolls over)

`last_connect_time_` will have to be set in `setup()` to some suitable value (like the time at setup)

jeromelaban replied:
Interesting indeed. ~40 days may not be particularly interesting either, all things considered. I could not find a 64 bits version of the counter, is there one to be used ?

OttoWinter commented on esphome/components/airthings_wave_plus/airthings_wave_plus.h:
```diff
@@ -0,0 +1,90 @@
+  };
+
+  struct WavePlusReadings {
+    unsigned char version;
+    unsigned char humidity;
+    unsigned char ambientLight;
+    unsigned char unused01;
+    unsigned short radon;
+    unsigned short radon_lt;
+    unsigned short temperature;
+    unsigned short pressure;
+    unsigned short co2;
+    unsigned short voc;
```
```suggestion
    uint8_t version;
    uint8_t humidity;
    uint8_t ambientLight;
    uint8_t unused01;
    uint16_t radon;
    uint16_t radon_lt;
    uint16_t temperature;
    uint16_t pressure;
    uint16_t co2;
    uint16_t voc;
```

See also the lint messages

OttoWinter commented on esphome/components/airthings_wave_plus/sensor.py:
```diff
@@ -0,0 +1,99 @@
+)
+
+ESP_PLATFORMS = [ESP_PLATFORM_ESP32]
+DEPENDENCIES = []
```
```suggestion
DEPENDENCIES = ["esp32_ble_tracker"]
```

I think ESP32 ble tracker is required for this component, no?

jeromelaban replied:
Not in this current implementation, but should it? The BLE initialization is performed entirely by this component.

OttoWinter commented on esphome/components/airthings_wave_plus/airthings_wave_plus.cpp:
```diff
@@ -0,0 +1,182 @@
+      new WavePlusClientCallbacks([this] { this->client_connected_(); }, [this] { this->client_disconnected_(); });
+  client_->setClientCallbacks(client_callbacks);
+
+  set_interval("connect", 10000, [this] { this->update_(); });
```
If this is an update interval, `PollingComponent` should be used instead (then the `update()` method is called periodically like here)

jeromelaban replied:
It really is an updated interval. I've updated the component to inherit from `PollingComponent` instead.

OttoWinter reviewed (commented):
Thanks for this contribution!

I took a look at the code, and found a couple of things that can be improved. Could you take a look? Thanks.

I don't know much about our BLE APIs, @jesserockz could you take a look at those parts?

jeromelaban committed:
chore: add set_address suggestion

jeromelaban committed:
chore: Adjustments from code review

jeromelaban committed:
chore: Refactor to polling component

jeromelaban committed:
chore: linter adjustments

jeromelaban committed:
chore: additional linter fixes

jesserockz reviewed (changes requested):
Basically the same comment as @buxtronix around the BLE client...

I am thinking that this is not compatible with `esp32_ble_tracker` as it reinitializes the BLE stack for its own use.

This would need to be updated to utilise the internal ESPHomes classes and not the Arduino Library BLE.

jeromelaban committed:
refactor: Move to BLEClientNode

This change delegates BLE managenement to BLEClientNode and BLEClient, instead of the integration initializating BLE itself.

jeromelaban commented:
> This would need to be updated to utilise the internal ESPHomes classes and not the Arduino Library BLE.

I've refactored the extension to use BLEClient / BLEClientNode as the base for handling BLE connectivity. The only trick here is that the BLE Client is built to stay connected and the airthings wave only allows for one connection at a time, preventing other apps from connection.

To cope for that, once values are read, the BLEClient is disabled until the next update invocation.

jeromelaban committed:
chore: Adjust for linter suggestions

jeromelaban committed:
chore: Adjust test for airthings

jeromelaban committed:
chore: Adjust for more linter suggestions

jeromelaban committed:
chore: Fix test2 validation

jesserockz commented on esphome/const.py:
```diff
@@ -834,6 +838,7 @@
 DEVICE_CLASS_TEMPERATURE = "temperature"
 DEVICE_CLASS_POWER_FACTOR = "power_factor"
 DEVICE_CLASS_PRESSURE = "pressure"
+DEVICE_CLASS_RADON = "radon"
```
Device classes have to exist in [Home Assistant](https://rc.home-assistant.io/integrations/sensor/) before they can be added to ESPHome.
If it does not exists in HA, then `DEVICE_CLASS_NONE` should be used, or it should be left out of the named args of `sensor_schema()` (Like I modified)

jeromelaban replied:
Thanks, I've removed class, that's good to know! That being said, I've been using it in home assistant with this for a while, and I have not noticed any warning or error message related to this. What are the consequences ?

jesserockz replied:
There are no consequences, but we don't want to introduce classes that do not exist in HA. You are able to request the class be added in HA (not sure of the process, perhaps see a past PR that does this)

jesserockz committed:
Make some config changes

jeromelaban committed:
chore: Additional removes of DEVICE_CLASS_RADON

jeromelaban committed:
chore: remove invalid added class to sensor

jesserockz committed:
Add state classes

XBeg9 commented:
will this be released soon?

Srki82 commented:
Hi,
Sorry for maybe a simple question but how do you enable the BLE Scanner to find the airthings MAC asdress? When I use the following code in my config.yaml file to find the airthings MAC:

**esp32_ble_tracker:
airthings_wave_plus:** 

I get an error for the row airthings_wave_plus: 
I have installed the latest beta 2021.9.0b1 but it is like airthings_wave_plus cannot be included into the code.
What am I missing?

jesserockz commented:
@Srki82 Please look at the example carefully https://next.esphome.io/components/sensor/airthings_ble.html#device-discovery

The discovery YAML key is not `airthings_wave_plus`

Srki82 commented:
@jesserockz  thanks! I followed the instructions and I managed to find the MAC adress and then I got the sensors inside Home Assistant. The only issue now is that all sensors shows the state "unknown". 
I assume its not communicating with the airthing plus unit for some reason. Any further tips?

jeromelaban commented:
> @jesserockz thanks! I followed the instructions and I managed to find the MAC adress and then I got the sensors inside Home Assistant. The only issue now is that all sensors shows the state "unknown".
> I assume its not communicating with the airthing plus unit for some reason. Any further tips?

If all the states are unknown, this means there's a communication failure somewhere. I'd say this PR this pr is not the right place to discuss this, and a new issue would be more interesting to provide additional information. [Can you open a new one](https://github.com/esphome/issues/issues) with debugging information from the ESP?
